### ISSUE TITLE
feat: make `shorebird` easier to use by developers in China

### ISSUE THREAD

**Issue by eseidel** [shorebirdtech/shorebird#435]
Update Sept 11, 2024:  We know of at least some developers using Shorebird successfully from China today.
Users should be able to download published patches through our CDN (without needing to do anything special):
https://shorebird.dev/blog/improved-patch-delivery/
However the updater client also talks to `api.shorebird.dev` which is hosted in Google Cloud and thus may not be accessible.

Developers may also still need to find a way to reach Google Cloud for using `shorebird` tooling such as using a VPN.

=====

The directions for using Shorebird in china include setting FLUTTER_STORAGE_BASE_URL:
https://docs.flutter.dev/community/china

This environment variable is ignored by `shorebird`, since `shorebird` itself uses `FLUTTER_STORAGE_BASE_URL` to pass `download.shorebird.dev` to `flutter`.

The larger issue for developers in China is that Shorebird currently uses Google Cloud as our infrastructure provider and Google Cloud is not reliably reachable across China.

We could implement a proxy within china for `download.shorebird.dev` and either then make it possible to use FLUTTER_STORAGE_BASE_URL to point to such, or some other environment variable.

However we would still need to make access to `api.shorebird.dev` more reliable for Shorebird to work for Developers without them needing to modify their network.

Regardless, Shorebird works seems to work for some users in china (the api.shorebird.dev request sometimes fails, but if it succeeds patch installs otherwise seem to work fine in China).   Developers who cannot reach api.shorebird.dev will still have trouble actually uploading the patches.

**Comment by xster** [comment]
I suspect it's not as simple as setting an alternate URL. The mirror URLs pointed to in https://docs.flutter.dev/community/china physically host the actual bytes to the Flutter engine etc. You'll need to find a partner (who can register a https://en.wikipedia.org/wiki/ICP_license) who can host the app bits being pushed to point the CLI to.

**Comment by AlexV525** [comment]
Hi guys, just noticed this issue when I was browsing new stuff. I can answer the questions above since I'm also a maintainer to flutter.cn .

> I'm not sure if download.shorebird.dev is accessible from China?

Currently, it's available. If you can provide an exact artifact destination then we can test if the connection is stable enough (download.shorebird.dev returns `Unrecognized artifact path`).

> I'm also not sure how we would tell download.shorebird.dev (possibly in the headers?) what proxy to proxy too?

About how our mirror works: We host our service with a local services provider, mirroring requests with the same path of googleapis, marking artifacts that need to be cached by CDN, and resolving them into buckets.

In order to make our mirror work with shorebird, we'll need another domain or subdomain for it, and create another service that is different from the Flutter one.

So to summarize this, I don't think we'll need a new mirror service for now since everything is working as expected. Let me know if I missed something.

**Comment by eseidel** [comment]
Thanks Alex.  That sounds like an excellent option.  I don't think we're quite ready to set this up yet, but if you reach out via email eric@, happy to discuss further when we get there!

**Comment by liu731** [comment]
another way is to support private deployment.

**Comment by eseidel** [comment]
We also just got a report that Google OAuth may not work in China (may not connect to oauth2.googleapis.com).  So we likely need to support https://github.com/shorebirdtech/shorebird/issues/335 too.

Recommended alternatives included WeChat or QQ OAuth.
Also just having Shorebird support its own accounts (e.g. with email auth).

**Comment by 956569686** [comment]
When will this problem be solved?

**Comment by eseidel** [comment]
We don't have a specific date.  We're quite busy with other customers at the moment and have not yet sat down to try and understand what work remains to be done for Shorebird to be usable within China.

**Comment by 956569686** [comment]
I hope the team can support Chinese developers as soon as possible. A large number of Chinese flutter developers hope to use this tool.

**Comment by ZhouC125** [comment]
Hoping for Shorebird support in China soon!

**Comment by 1leontao** [comment]
Hoping for Shorebird support in China soon!

**Comment by li8607** [comment]
Hoping for Shorebird support in China soon!

**Comment by nailongyeshilong** [comment]
> the directions for using Shorebird in china include setting FLUTTER_STORAGE_BASE_URL: https://docs.flutter.dev/community/china
> 
> Which I believe will not work with our current setup.
> 
> I'm not entirely sure how we will make it work as:
> 
> 1. I'm not sure if download.shorebird.dev is accessible from China?
> 2. I'm also not sure how we would tell download.shorebird.dev (possibly in the headers?) what proxy to proxy too?
> 
> Or maybe we'd work with a local china proxy to have both the shorebird and Flutter artifacts, once #434 is fixed?
> 
> This could also go away if we fix: #434. As it's possible the way we hook in does not involve setting FLUTTER_STORAGE_BASE_URL but rather modifying `flutter` to teach it about our additional artifact type.

If someone in Chinese mainland can't login with 'shorebird login', maybe you need set proxy to your terminal, just input follow command(if you use clash in windows os):
set http_proxy=http://127.0.0.1:7890/
set https_proxy=https://127.0.0.1:7890/
shorbird login

筒子们如果发现开了全局代理之后，用shorebird login命令也登陆不了的话，试下给终端把代理设置上再去登陆！

**Comment by 956569686** [comment]
set http_proxy=http://127.0.0.1:7890/
set https_proxy=https://127.0.0.1:7890/
这种方式可行吗

**Comment by okFancy** [comment]
Hello, Chinese developers hope that Shorebird can support usage in China.

**Comment by franticn** [comment]
I think you can solve this problem by mirroring. The most common use in China is TUNA, Tsinghua's own site mirroring a wide range of open source software, These include FLUTTER_STORAGE_BASE_URL and PUB_HOSTED_URL, which can be reached at support@tuna.tsinghua.edu.cn

```bash
export FLUTTER_STORAGE_BASE_URL="https://mirrors.tuna.tsinghua.edu.cn/flutter"
export PUB_HOSTED_URL="https://mirrors.tuna.tsinghua.edu.cn/dart-pub"
```

* https://mirrors.tuna.tsinghua.edu.cn/ 
* https://mirrors.tuna.tsinghua.edu.cn/help/flutter/

**Comment by dongjianxiong** [comment]
Hoping for Shorebird support in China soon! Thank you very much!

**Comment by eseidel** [comment]
We have quite a few customers from China.  So this appears to work "well enough" for the moment.  I think when we moved our distribution to a CDN that will help customers in china (since their users will get updates more reliably).

**Comment by Heilum** [comment]
A simple solution: just adding a optional http-proxy-string parameter to ShorebirdCodePush's constructor and let developers from China to use shorebird_code_push package and manually do code-pushing!

**Comment by eseidel** [comment]
Update: We've updated our infrastructure to use CloudFlare CDN, which as far as we can tell is more reliably reachable across China:
https://shorebird.dev/blog/improved-patch-delivery/

We still make direct connections to Google Cloud as part of using `shorebird` on a developer machine.  So developers in parts of the world that do not have reliable access to Google Cloud will still need to use a VPN or similar access mechanism, but users should now have more reliable access to download patches published with Shorebird.

**Comment by huzhiju** [comment]
I uploaded the patch, but Chinese users are unable to download it

**Comment by huzhiju** [comment]
Can we open the server API and build our own download service

**Comment by bryanoltman** [comment]
> Can we open the server API and build our own download service

This is captured in https://github.com/shorebirdtech/shorebird/issues/485. The short answer is that we do not currently have plans to support this.

**Comment by wubaibin** [comment]
In fact, this is not the case, and patch developers can release it by climbing the wall. But Chinese users can't download it normally

**Comment by rockingdice** [comment]
@eseidel  There’s one thing I don’t quite understand: Can I assume that if a user in China can obtain the result of a network request through the checkForUpdate method of the shorebird_code_push plugin, they would also be able to download the patch using the same network environment?

**Comment by felangel** [comment]
> [@eseidel](https://github.com/eseidel) There’s one thing I don’t quite understand: Can I assume that if a user in China can obtain the result of a network request through the checkForUpdate method of the shorebird_code_push plugin, they would also be able to download the patch using the same network environment?

Currently the device makes a request to https://api.shorebird.dev to check if an update is available and if available, the update is downloaded from https://cdn.shorebird.cloud.

**Comment by rockingdice** [comment]
> > [@eseidel](https://github.com/eseidel) There’s one thing I don’t quite understand: Can I assume that if a user in China can obtain the result of a network request through the checkForUpdate method of the shorebird_code_push plugin, they would also be able to download the patch using the same network environment?
> 
> Currently the device makes a request to https://api.shorebird.dev to check if an update is available and if available, the update is downloaded from https://cdn.shorebird.cloud.

Thank you for your quick reply! It seems the issue is with the second URL.

![Image](https://github.com/user-attachments/assets/9d668e23-dd98-4bc3-9d16-f2050a3cdcb2)

You can test the accessibility of the service in China using this tool: https://ping.chinaz.com/cdn.shorebird.cloud.
I just realized that you optimized the network for China last year, as mentioned in this blog post: https://shorebird.dev/blog/improved-patch-delivery/. The improvements described there convinced me to try your service, but after testing it myself, it seems that the issue still persists. Could this be fixed?
I have around 500,000 users in my apps, so it's really important to have a stable service.

**Comment by rockingdice** [comment]
Could you consider using a Chinese local cloud service for the CDN if the users are in China? It would be beneficial to have a local CDN for better stability and speed.

**Comment by felangel** [comment]
Using the ping tool you linked, this is what I see as of now:
<img width="1210" alt="Image" src="https://github.com/user-attachments/assets/662d39b1-fa1c-4c38-bddc-ae10352ec585" />

We do plan to improve this in the future -- Cloudflare also has [solutions to provide better availability in China](https://developers.cloudflare.com/china-network/#:~:text=The%20data%20centers%20cover%20most,performing%20content%20delivery%20in%20China.) but we're currently working on some higher priority items and we'll get to this as soon as we can 👍

**Comment by rockingdice** [comment]
@felangel You can hover over each province to see if the local operators are accessible. Even in Beijing two of them are not able to connect:

![Image](https://github.com/user-attachments/assets/111891a2-8a16-42aa-8cbc-c69def57e3ff)

Thank you and cheers!

**Comment by leiyangyou** [comment]
A solution could be to have a mirroring proxy in China that has stable access to cloudflare and a proxy to the API end

Then use a custom version of the push client

**Comment by fengnudexiaoniao** [comment]
作为开发者我们用代理就是解决开发环境的问题，目前最大的问题就是用户设备端热更新的下载，下载补丁下载不下来，咱们的补丁发布的服务器如果可以支持国内就好了。

**Comment by guoshan-yang** [comment]
希望能够支持中国市场

**Comment by chenglu** [comment]
@felangel Cloudflare does provide services for better availability in China, but it's not that easy to set up in practice.

To make it work, the domain needs to go through ICP filing and be hosted by a mainland provider like Alibaba, Tencent Cloud, which means you will need to transfer the domain shorebird.cloud from Cloudflare to Aliyun (not the global version of Alibaba Cloud — the domain service provider must be located in mainland China).

Additionally, there are occasional manual reviews by regulatory authorities to ensure the site doesn’t contain any "restricted content", and someone must be available to answer those compliance calls. And I’m sure the Flutter team has already thought this through and come up with the [smartest approach](https://docs.flutter.dev/community/china) ;-)

I've been following this thread for a while. I truly hope the developers in the China community can fully enjoy and experience your amazing service. So I’d love to share some of my experience here, hopefully help you avoid unnecessary detours. 

cc/ @eseidel

**Comment by chenkaiHere** [comment]
It seems to be unavailable in mainland China at the moment. I hope to improve the level of the task of relocating Cloudflare.

<img width="799" height="843" alt="Image" src="https://github.com/user-attachments/assets/b0c17d6e-c9c1-414a-b7f4-cd9c7222fd14" />

**Comment by tomarra** [comment]
> It seems to be unavailable in mainland China at the moment. I hope to improve the level of the task of relocating Cloudflare.

Unfortunately it's a bit of a cat & mouse game that changes daily. I just reran the same query and while it's not all green it's much closer, https://tool.chinaz.com/dns/cdn.shorebird.cloud

<img width="1204" height="789" alt="Image" src="https://github.com/user-attachments/assets/2d561036-d9bc-45d3-954a-557d3789ea65" />

Overall I would love to be able to fully support the China market but as a lot of the comments on this ticket have said it's a tricky setup to build & maintain. The hope is as we continue to grow it becomes easier to support but we don't have a solid timeline for that right now.

**Comment by chenglu** [comment]
@tomarra you should use the ping tool instead of the dns checker tool ;-D
https://ping.chinaz.com/cdn.shorebird.cloud

**Comment by azhaorz** [comment]
Is it possible to put “cdn.shorebird.cloud” into “shorebird.yaml” as a configurable item so that I can handle the proxy issue myself.

**Comment by realpg** [comment]
> Update: We've updated our infrastructure to use CloudFlare CDN, which as far as we can tell is more reliably reachable across China: https://shorebird.dev/blog/improved-patch-delivery/
> 
> We still make direct connections to Google Cloud as part of using `shorebird` on a developer machine. So developers in parts of the world that do not have reliable access to Google Cloud will still need to use a VPN or similar access mechanism, but users should now have more reliable access to download patches published with Shorebird.

Cloudflare is very slow in China, sometimes completely unreachable for some chinese carriers.

Chinese developers will find their own ways to speed up their access. However, app users who need updates cannot obtain the update packages normally on their own. This means that in essence, Shorebird cannot be used in the production environment in mainland China at all.

I've also read many issues of this project. Many people have mentioned this problem and proposed solutions such as self - hosting. I've read historical discussion threads like #485 and understand and respect the Shorebird team's decision not to support self - hosting.

There are one billion people in mainland China, which is a huge market. Chinese developers understand the convenience of the Shorebird project and hope to use it. Currently, it is completely impossible to use Shorebird hot patch in the production environment in mainland China.

I'd like to ask if your team has any plans, or hopes to explore the mainland Chinese market by finding a partner company in China?

**Comment by eseidel** [comment]
We have no current plans to find a partner company in China.  Our lawyers have previously advised us against such until we're significantly larger in size.

**Comment by CyJaySong** [comment]
Reverse proxy scheme

https://github.com/shorebirdtech/shorebird/issues/3358#issuecomment-3515717234

**Comment by eseidel** [comment]
We've been very busy getting the first pass CI product launched, but are now back to looking at core platform-level issues like this one.  This particular issue is not scheduled until at least Q1 next year, but my current thinking is:

1. Reduce the number of required endpoints that your app needs to reach to talk to Shorebird.  Currently it's 3: cloudflare, GCP cloud run and GCP cloud storage.  We should be able to reduce that to just 1 (cloudflare).
2. Then we can separately explore what regional-specific hosting could cost with Cloudflare.  My understanding is they do offer CN-specific hosting for example. https://www.cloudflare.com/application-services/products/china-network/ and we have had several enterprise customers reach out from within China, potentially sufficient enough for us to explore such a china-specific cost in 2026.

Until then, there has been exploration of individuals setting up their own proxies as @cyjaysong links to in https://github.com/shorebirdtech/shorebird/issues/435#issuecomment-3522469829

**Comment by qcks** [comment]
Please don't support China, the people here are tired enough, people die from overwork every year, and the stable version release route is the way of life of normal people

**Comment by CyJaySong** [comment]
@qcks  你的发言偏激，显得非常的没修养。难道其他国家没人因为劳累而猝死的吗？难道因为有中国开发者因过度工作而死亡，其他中国开发者不配享受便利的工具吗？你要发泄你的情绪请移步996.icu

**Comment by eseidel** [comment]
We'e meeting with a cloud provider today about expanding our reachability within China.  Separately we're hiring a backend engineer, for whom this will be a priority.  I expect we'll have more updates here within another month or so.

**Comment by dongjianxiong** [comment]
我怎么惹你了？

---- Replied Message ----
| From | ***@***.***> |
| Date | 01/29/2026 10:38 |
| To | shorebirdtech/shorebird ***@***.***> |
| Cc | 董剑雄 ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [shorebirdtech/shorebird] feat: make `shorebird` easier to use by developers in China (Issue #435) |
CyJaySong left a comment (shorebirdtech/shorebird#435)

@qcks 你是一个极端而又偏激的人。难道其他国家没人因为劳累而猝死的吗？难道因为有人因过度工作而死亡，其他中国开发者不配享受便利的工具吗？你要发泄你的情绪请移步996.icu

**Comment by guopengSN** [comment]
> We'e meeting with a cloud provider today about expanding our reachability within China. Separately we're hiring a backend engineer, for whom this will be a priority. I expect we'll have more updates here within another month or so.

How's it going?

**Comment by eseidel** [comment]
We got a quote from the provider and we have a new engineer hired.  We're now waiting for our backend engineer to start.  I don't expect we'll have an update here for at least another month.

**Comment by HaniHorani** [comment]
"I'm building an app for users in Syria. All Shorebird URLs are blocked there. Happy to test any proxy or mirror solution if you build one. This would be very meaningful for real users on the ground."

**Comment by hl1716** [comment]
> 我们已经从供应商那里拿到了报价，也聘请了一位新的工程师。现在我们正在等待后端工程师入职。我估计至少一个月内不会有新的进展。

？？

**Comment by eseidel** [comment]
We got the quote.  The backend eng started a few weeks ago. Just been busy with other tasks first.

**Comment by Zensonaton** [comment]
Users from Russia aren't able to use Shorebird without VPN as well: technically, CDN works and updates are being downloaded, but their speed is limited to ~4kbps by ISPs (probably due to DPI).
I've tested on 2 ISPs, asked my friends to test download speed, and the only way to get patches downloaded properly is to turn on VPN.

**Comment by xiaoxiaokanhu** [comment]
@eseidel What's the progress of the China compatibility update? Your reply is appreciated.

**Comment by eseidel** [comment]
We remain busy with other tasks.  We've done some of the refactoring work necessary to expose Shorebird directly via JD Cloud, but I still do not have an exact timeline to offer on this work.

**Comment by xiaoxiaokanhu** [comment]
@eseidel 
Thanks for the update.

China compatibility remains a very important feature for many developers. It's encouraging to hear that the JD Cloud-related refactoring is underway.

Looking forward to any future updates when there's more progress to share.

**Comment by lei-6-wang** [comment]
> We remain busy with other tasks. We've done some of the refactoring work necessary to expose Shorebird directly via JD Cloud, but I still do not have an exact timeline to offer on this work.

@eseidel  is there any progress？

**Comment by eseidel** [comment]
We remain engineering constrained.  (Engineers busy with other issues.)  No progress to report at this time.

**Comment by Mitsuha** [comment]
> We remain engineering constrained. (Engineers busy with other issues.) No progress to report at this time.

I’m a developer based in China. My users are still unable to update via Shorebird. Would it be possible to prioritize setting up a China-based CDN for the download server? While other networking issues on the development side are tolerable, the inability for users to update through Shorebird renders the service completely unusable. If any ICP-related requirements (such as obtaining an ICP license in China) are needed, my company would be happy to assist (for free)

**Comment by CyJaySong** [comment]
@Mitsuha 我使用服务器做api反代和国内的CDN中转成功实现了顺畅的热更新，这并不是什么难事，我很乐意提供帮助。

**Comment by Mitsuha** [comment]
> [@Mitsuha](https://github.com/Mitsuha) 我使用服务器做api反代和国内的CDN中转成功实现了顺畅的热更新，这并不是什么难事，我很乐意提供帮助。

谢谢，可以在你的博客留下你的联系方式吗，想请教一些关键问题

**Comment by CyJaySong** [comment]
> > [@Mitsuha](https://github.com/Mitsuha) 我使用服务器做api反代和国内的CDN中转成功实现了顺畅的热更新，这并不是什么难事，我很乐意提供帮助。
> 
> 谢谢，可以在你的博客留下你的联系方式吗，想请教一些关键问题

博客首页和我的Github个人信息显示了邮箱地址

**Comment by huangjinpeng** [comment]
@CyJaySong 你好，给你发了邮件希望得到回复

**Comment by CyJaySong** [comment]
> [@CyJaySong](https://github.com/CyJaySong) 你好，给你发了邮件希望得到回复

@huangjinpeng 已回邮件

**Comment by cenumi** [comment]
> [@Mitsuha](https://github.com/Mitsuha) 我使用服务器做api反代和国内的CDN中转成功实现了顺畅的热更新，这并不是什么难事，我很乐意提供帮助。

不违规吗这个

**Comment by cdoky** [comment]
提供一个示例配置
```
# api.xxx.cn — Shorebird patch API reverse proxy

server {
    listen 127.0.0.1:8443 ssl;
    listen [::1]:8443 ssl;
    server_name api.xxx.cn;

    ssl_certificate     /etc/letsencrypt/live/api.xxx.cn/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/api.xxx.cn/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
    include /etc/nginx/snippets/ssl-security.conf;

    location / {
        proxy_pass https://api.shorebird.dev;
        proxy_set_header Host api.shorebird.dev;
        proxy_ssl_server_name on;
        proxy_ssl_name api.shorebird.dev;

        proxy_http_version 1.1;
        proxy_connect_timeout 10s;
        proxy_read_timeout 30s;

        # Patch checks and event reports are dynamic POST requests; never cache.
        proxy_no_cache 1;
        proxy_cache_bypass 1;
        add_header Cache-Control "no-store" always;
    }
}

server {
    listen 80;
    listen [::]:80;
    server_name api.xxx.cn;

    location /.well-known/acme-challenge/ {
        root /var/www/certbot;
    }

    location / {
        return 301 https://$host$request_uri;
    }
}

```

**Comment by CyJaySong** [comment]
@cdoky 不要只转发/api/v1/patches/，会有其他问题，请将示例改为全量转发，以免引来不必要的麻烦，有些事情使用者自己悟就行。

<img width="32" height="32" alt="Image" src="https://github.com/user-attachments/assets/8678b613-2cb5-475f-8cd1-abc5b756e868" />
